### PR TITLE
network/VR: fix dhcp/password/metadata issues on shared networks with multiple subnets

### LINKED ISSUE #4943: data-server breaks VM root password in multi-subnet flat network
##### ISSUE TYPE

 * Bug Report
 
##### COMPONENT NAME

~~~
VM password (for cloud-init)
~~~

##### CLOUDSTACK VERSION

~~~
4.15
~~~

##### CONFIGURATION

Basiz zone with multiple subnets/CIDR

##### OS / ENVIRONMENT

VMware 6.5 (limited access to environment, not "mine")

##### SUMMARY

The VM password feature of cloud-init Cloudstack datasource attempts retrieval of the password from "data-server" host if it resolves. 
What happens in a multi-CIDR basic zone is that the passwords for the VMs are stored in separate files, according to their subnet, eg:
/var/cache/cloud/passwords-subnet1IP
/var/cache/cloud/passwords-subnet2IP and so on
The problem arises when Cloudstack adds the "data-server" hostname in /etc/hosts for dnsmasq to pick up, but because it adds it with subnet1IP, then requests for a password from a machine in subnet2 or subnet3 will go unanswered.

The Cloudstack datasource does not handle this gracefully, it just fails setting a password instead of detecting there was no valid response and try to also ask the "dhcp_identifier" host.

I suggest as a quick workaround to add a button somewhere to disable the data-server feature altogether. It seems to be Cloudstack specific, grepping for it in other data sources yielded nothing. 

On the VMs right now we have to add "0.0.0.0 data-server" in /etc/hosts to make the root password work. I guess editing the /etc/hosts on the VR and remove the data-server entry might also work, until the next restart.

##### STEPS TO REPRODUCE

~~~
Deploy basic or adv+SG zone with multiple CIDRs, 
when asking for a password from a VM not in the primary CIDR this will fail, password will not be set.
~~~

##### EXPECTED RESULTS

~~~
data-server hostname should return a valid password regardless of which CIDR the request comes from.
~~~

##### ACTUAL RESULTS

~~~
Valid password not served.
~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #4943

This PR contains two changes
(1) assign ip in new subnet as alias of guest IP in first subnet.
If there are free IPs in first subnet, cloudstack acquires a IP from them as alias, which is obviously wrong(dhcp does not work).
Java code changes in this PR fixes it.

(2) add missing iptables rules for port 80(apache2/metadata) and port 8080 (password server).
The rules are added in python but not applied in VR, which cause password server and metadata/userdata not work in new subnet. 
Python code changes in this PR fixes it.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### CODE REVIEW

File: systemvm/debian/opt/cloud/bin/configure.py [hunk 1]
```diff
@@ -1070,8 +1070,8 @@ def main(argv):
     config.address().process()
 
     databag_map = OrderedDict([("guest_network",     {"process_iptables": True,  "executor": []}),
-                               ("vm_password",       {"process_iptables": False, "executor": [CsPassword("vmpassword", config)]}),
-                               ("vm_metadata",       {"process_iptables": False, "executor": [CsVmMetadata('vmdata', config)]}),
+                               ("vm_password",       {"process_iptables": True, "executor": [CsPassword("vmpassword", config)]}),
+                               ("vm_metadata",       {"process_iptables": True, "executor": [CsVmMetadata('vmdata', config)]}),
```
Comment (rhtyd) [inline systemvm/debian/opt/cloud/bin/configure.py:1074]: @weizhouapache can you explain how does this fix the issue? (I've no idea how the key is used, however ignore my comment if the PR is still WIP)

**Comment by weizhouapache** [comment]
@NuxRo @ccycv @Spaceman1984  @rhtyd 
this PR is ready for review and testing !

**Comment by rhtyd** [comment]
@weizhouapache can you kick packaging and smoketest using blueorangutan?

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by ccycv** [comment]
> @NuxRo @ccycv @Spaceman1984 @rhtyd
> this PR is ready for review and testing !

Great, once is released I will update the ACS and test.

**Comment by weizhouapache** [comment]
> Trillian test result (tid-659)
> Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
> Total time taken: 57177 seconds
> Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5013-t659-kvm-centos7.zip
> Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
> Intermittent failure detected: /marvin/tests/smoke/test_routers_network_ops.py
> Intermittent failure detected: /marvin/tests/smoke/test_volumes.py
> Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
> Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
> Smoke tests completed. 85 look OK, 2 have error(s)
> Only failed tests results shown below:
> 
> Test	Result	Time (s)	Test File
> test_02_isolate_network_FW_PF_default_routes_egress_false	`Failure`	272.24	test_routers_network_ops.py
> test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true	`Failure`	310.15	test_routers_network_ops.py
> test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false	`Failure`	307.80	test_routers_network_ops.py
> test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers	`Failure`	423.80	test_vpc_redundant.py

@rhtyd I will address test errors above.

**Comment by weizhouapache** [comment]
> > @NuxRo @ccycv @Spaceman1984 @rhtyd
> > this PR is ready for review and testing !
> 
> Great, once is released I will update the ACS and test.

@ccycv there is a easy way to test it.
(1) log into existing VR
(2) open /opt/cloud/bin/configure.py , add new line 
```
                               ("ip_aliases",        {"process_iptables": True,  "executor": []}),
```
(3) stop/start a vm, or create a new vm. the rules in VR will be reprogrammed.
(4) password and userdata/meta-data should work.

**Comment by Spaceman1984** [comment]
I'll test

**Comment by Spaceman1984** [comment]
I created a basic network with multiple IP ranges with different CIDRs and I am not able to create more instances after the first range is exhausted. Not sure if this PR is meant to address this issue.

**Comment by weizhouapache** [comment]
> I created a basic network with multiple IP ranges with different CIDRs and I am not able to create more instances after the first range is exhausted. Not sure if this PR is meant to address this issue.

@Spaceman1984 which branch did you test ? can you share some error logs ?

**Comment by ccycv** [comment]
> > I created a basic network with multiple IP ranges with different CIDRs and I am not able to create more instances after the first range is exhausted. Not sure if this PR is meant to address this issue.
> 
> @Spaceman1984 which branch did you test ? can you share some error logs ?

Are you sure that the VR IP is not used by any of those VMs?

**Comment by Spaceman1984** [comment]
The VR addresses are
```
Guest IP Address
192.168.43.74
Link Local IP Address
192.168.43.40

```
Ranges are

<html>
<body>


IPv4 Start IP | IPv4 End IP | IPv6 Start IP | IPv6 End IP | Gateway | Netmask |  
-- | -- | -- | -- | -- | -- | --
192.168.43.71 | 192.168.43.75 |   |   | 192.168.43.1 | 255.255.255.0 |  
10.2.1.31 | 10.2.1.40 |   |   | 10.2.1.1 | 255.255.255.0


</body>
</html>

**Comment by weizhouapache** [comment]
> The VR addresses are
> 
> ```
> Guest IP Address
> 192.168.43.74
> Link Local IP Address
> 192.168.43.40
> ```
> 
> Ranges are
> 
> IPv4 Start IP	IPv4 End IP	IPv6 Start IP	IPv6 End IP	Gateway	Netmask	 
> 192.168.43.71	192.168.43.75	 	 	192.168.43.1	255.255.255.0	 
> 10.2.1.31	10.2.1.40	 	 	10.2.1.1	255.255.255.0	
> VMS are using 192.168.43.75 and 192.168.43.73

Investigated with @Spaceman1984 , this looks like a primate UI bug.
podid is passed in legacy UI, but not passed in primate UI, when create vlan ip range.

**Comment by Spaceman1984** [comment]
Tested LGTM

**Review by Spaceman1984 — APPROVED** [review]

**Review by rhtyd — APPROVED** [review]

**Pull request merged**